skfp/utils/splitters.py

```diff
@@ -0,0 +1,630 @@
+        provided SMILES
+
+    References
+    ----------
+    .. [1] `Bemis, G. W., & Murcko, M. A.
+        "The properties of known drugs. 1. Molecular frameworks."
+        Journal of Medicinal Chemistry, 39(15), 2887-2893.
+        https://www.researchgate.net/publication/14493474_The_Properties_of_Known_Drugs_1_Molecular_Frameworks`_
+
+    .. [2] `Z. Wu, B. Ramsundar, E. N. Feinberg, J. Gomes, C. Geniesse, A. S. Pappu, K. Leswing, V. Pande
+        "MoleculeNet: A Benchmark for Molecular Machine Learning."
+        Chemical Science, 9(2), 513-530.
+        https://www.researchgate.net/publication/314182452_MoleculeNet_A_Benchmark_for_Molecular_Machine_Learning`_
+
+    """
+    _validate_split_sizes(train_size, test_size)
+    train_size, valid_size, test_size = _split_size(train_size, valid_size, test_size)
+
+    scaffolds = _create_scaffolds(list(data), include_chirality)
+    scaffold_sets = sorted(scaffolds.values(), key=len)
+
+    train_ids: list[int] = []
+    valid_ids: list[int] = []
+    test_ids: list[int] = []
+
+    train_ids, valid_ids, test_ids = _split_ids_three_sets(
+        scaffold_sets, len(data), test_size, valid_size
+    )
+
+    train_subset: list[Any] = []
+    valid_subset: list[Any] = []
+    test_subset: list[Any] = []
+
+    if return_indices:
+        train_subset = train_ids
+        valid_subset = valid_ids
+        test_subset = test_ids
+    else:
+        train_subset = _get_data_from_indices(data, train_ids)
+        valid_subset = _get_data_from_indices(data, valid_ids)
+        test_subset = _get_data_from_indices(data, test_ids)
+
+    _check_subsets(train_subset, test_subset)
+
+    if len(valid_subset) == 0:
+        warnings.warn(
+            "Warning: Valid subset is empty. Consider using scaffold_train_test_split instead."
+        )
+
+    additional_data_split: list[Sequence[Any]] = []
+
+    additional_data_split = (
+        _split_additional_data(list(additional_data), [train_ids, valid_ids, test_ids])
+        if additional_data
+        else []
+    )
+
+    if additional_data:
+        return train_subset, valid_subset, test_subset, additional_data_split
+    else:
+        return train_subset, valid_subset, test_subset
+
+
+def _calculate_missing_sizes(
+    train_size: Optional[float],
+    valid_size: Optional[float],
+    test_size: Optional[float],
+) -> tuple[float, float, float]:
+    """
+    Calculate the missing sizes for train, validation, and test sets if they are not provided.
+
+    Parameters
+    ----------
+    train_size : float or None
+        The proportion of the data to use for training.
+    valid_size : float or None
+        The proportion of the data to use for validation.
+    test_size : float or None
+        The proportion of the data to use for testing.
+
+    Returns
+    -------
+    tuple
+        A tuple containing the proportions for training, validation, and testing sets.
+    """
```

Comment:
For internal functions, we don't keep full docstrings, since their maintenance would be too much trouble. Just the main description is enough, without `Parameters` and `Returns`. This applies to all internal functions, with underscore at the beginning.